tests/test_migration.py

```diff
@@ -0,0 +1,684 @@
+        with cities.open(encoding="utf-8", newline="") as file:
+            reader = csv.reader(file)
+            cls.header = next(reader)
+            cls.city_data = [transmogrify(row) for row in reader]
+
+        top_ten = cls.city_data[0:10]
+        cls.pops = np.array([city.pop for city in top_ten])
+        cls.distances = np.array([[distance(city1.lat, city1.long, city2.lat, city2.long) for city2 in top_ten] for city1 in top_ten])
+
+        return
+
+    def test_setup(self):
+        """Test the setup of the test class."""
+        assert self.header == [
+            "City",
+            "State",
+            "2023 Estimate",
+            "2020 Census",
+            "Change",
+            "mi^2",
+            "km^2",
+            "density/mi^2",
+            "density/km^2",
+            "latitude",
+            "longitude",
+        ], print(f"us-cities.csv header: {self.header} doesn't match expected header")
+        assert len(self.city_data) == 336, print(f"us-cities.csv has {len(self.city_data)} rows, expected 336")
+        assert self.city_data[0] == City(name="New York", pop=8258035, lat=40.66, long=73.94), print(
+            f"{self.city_data[0].name=} != 'New York'"
+        )
+        assert self.city_data[-1] == City(name="Davenport", pop=100354, lat=41.56, long=90.60), print(
+            f"{self.city_data[-1].name=} != 'Davenport'"
+        )
+
+        assert len(self.pops) == 10, print(f"self.pops has {len(self.pops)} elements, expected 10")
+        assert self.distances.shape == (10, 10), print(f"self.distances has shape {self.distances.shape}, expected (10, 10)")
+        assert np.all(self.distances.diagonal() == 0), print("self.distances.diagonal() != 0")
+        assert np.all(self.distances == self.distances.T), print("self.distances is not symmetric")
+
+        assert np.isclose(self.distances[0, 1], 3957.13675, atol=0.00001), print(
+            f"New York to Los Angeles distance is {self.distances[0, 1]}, expected 3957.13675"
+        )
+        assert np.isclose(self.distances[0, 9], 1342.25107, atol=0.00001), print(
+            f"New York to Jacksonville distance is {self.distances[0, 9]}, expected 1342.25107"
+        )
+
+        return
+
+    def test_gravity_model(self):
+        """Test the gravity model migration function without maximum."""
+        (k, a, b, c) = (0.1, 0.5, 1.0, 2.0)
+        network = gravity(self.pops, self.distances, k=k, a=a, b=b, c=c)
+        for _ in range(10):  # check 10 random values
+            i = np.random.randint(0, 10)
+            j = np.random.randint(0, 10)
+            if i != j:
+                dist = k * (self.pops[i] ** a) * (self.pops[j] ** b) * self.distances[i, j] ** (-1 * c)
+                assert np.isclose(network[i, j], dist), print(f"network[{i}, {j}] = {network[i, j]}, expected {dist=}")
+            else:
+                assert network[i, j] == 0, print(f"network[{i}, {j}] = {network[i, j]}, expected 0")
```

Comment:
This isn't getting called all the time (∴ not tested) due to my choice to select a few random values.

I think we should either go ahead and test _all_ the values (shouldn't take long) or have a separate check,
```python
assert np.all(network.diagonal == 0), "All diagonal values should be 0."
```